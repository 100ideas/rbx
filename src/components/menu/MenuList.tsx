import { cx } from "emotion";
import React from "react";

import Element from "components/element";
import { ModifierProps } from "modifiers";
import MenuListItem from "./MenuListItem";

export type MenuListModifierProps = Partial<{
  title: string;
}>;

export type MenuListProps = ModifierProps &
  MenuListModifierProps &
  Partial<Omit<React.ComponentPropsWithoutRef<"ul">, "unselectable">>;

interface MenuList extends React.ForwardRefExoticComponent<MenuListProps> {
  Item: typeof MenuListItem;
}

const MenuList: Partial<MenuList> = React.forwardRef<
  HTMLUListElement,
  MenuListProps
>(({ className, title, ...props }, ref) => (
  <React.Fragment>
    {title && <p className="menu-label">{title}</p>}
    <Element
      ref={ref}
      renderAs="ul"
      className={cx("menu-list", className)}
      {...props}
    />
  </React.Fragment>
));

MenuList.Item = MenuListItem;

export default MenuList as MenuList;
