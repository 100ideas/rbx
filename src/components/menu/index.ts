import "./menu.module.sass";

export { Menu } from "./menu";
