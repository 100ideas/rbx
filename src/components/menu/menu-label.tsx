import classNames from "classnames";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type MenuLabelProps = HelpersProps;

export const MenuLabel = forwardRefAs<HTMLParagraphElement, MenuLabelProps>(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("menu-label", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "p" },
);

MenuLabel.displayName = "Menu.Label";
