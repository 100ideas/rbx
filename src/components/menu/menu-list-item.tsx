import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { MenuList } from "./menu-list";

export type MenuListItemModifierProps = Partial<{
  active: boolean;
}>;

export type MenuListItemProps = Prefer<
  ModifierProps & MenuListItemModifierProps,
  React.HTMLAttributes<HTMLAnchorElement>
>;

export const MenuListItem = React.forwardRef<HTMLLIElement, MenuListItemProps>(
  (props, ref) => {
    const { active, children, className, ...rest } = props;

    if (typeof children === "string") {
      return (
        <li ref={ref}>
          <Element<"a">
            as="a"
            className={cx(className, { "is-active": active })}
            children={children}
            {...rest}
          />
        </li>
      );
    }

    if (React.Children.only(children).type === MenuList) {
      const child = React.Children.only(children);
      return (
        <li ref={ref}>
          <Element<"a">
            as="a"
            className={cx(className, { "is-active": active })}
            children={child.props.title}
            {...rest}
          />
          {React.cloneElement(child, { title: null })}
        </li>
      );
    }

    return <li ref={ref}>{children}</li>;
  },
);

MenuListItem.defaultProps = Object.assign(
  {
    active: false,
    children: null,
  },
  MenuListItem.defaultProps,
);
