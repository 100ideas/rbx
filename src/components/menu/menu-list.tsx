import classNames from "classnames";
import React from "react";

import { forwardRefAs } from "@/base";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { MenuListItem } from "./menu-list-item";

export type MenuListModifierProps = Partial<{ className: string }>;

export type MenuListProps = ModifierProps & MenuListModifierProps;

export const MenuList = Object.assign(
  forwardRefAs<MenuListProps, "ul">(
    (props, ref) => {
      const { as, ...rest } = transformModifiers(props);
      rest.className = classNames("menu-list", rest.className);
      return React.createElement(as!, { ref, ...rest });
    },
    { as: "ul" },
  ),
  {
    Item: MenuListItem,
  },
);
