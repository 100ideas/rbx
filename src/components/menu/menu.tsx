import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";
import { MenuList } from "./menu-list";

export type MenuProps = ModifierProps;

export const Menu = Object.assign(
  asExoticComponent<MenuProps, "aside">((props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("menu", rest.className);
    return React.createElement(as!, { ref, ...rest });
  }, "aside"),
  { List: MenuList },
);
