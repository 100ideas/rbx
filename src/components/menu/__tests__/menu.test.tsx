import React from "react";
import renderer from "react-test-renderer";

import { Menu } from "../menu";

describe("Menu component", () => {
  it("should Exist", () => {
    expect(Menu).toMatchSnapshot();
  });

  it("should have menu classnames", () => {
    const component = renderer.create(
      <Menu>
        <Menu.List title="General">
          <Menu.List.Item>Dashboard</Menu.List.Item>
          <Menu.List.Item>Customer</Menu.List.Item>
        </Menu.List>
      </Menu>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat Menu.List to display as submenus", () => {
    const component = renderer.create(
      <Menu>
        <Menu.List title="General">
          <Menu.List.Item>Dashboard</Menu.List.Item>
          <Menu.List.Item>
            <Menu.List.Item active>
              <Menu.List title="Manage Your Team">
                <Menu.List.Item>Members</Menu.List.Item>
                <Menu.List.Item active>Plugins</Menu.List.Item>
                <Menu.List.Item>Add a member</Menu.List.Item>
              </Menu.List>
            </Menu.List.Item>
          </Menu.List.Item>
        </Menu.List>
      </Menu>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render custom item child", () => {
    const component = renderer.create(
      <Menu>
        <Menu.List title="General">
          <Menu.List.Item>
            <p>Custom children 1</p>
          </Menu.List.Item>
          <Menu.List.Item>
            <a>Custom children 2</a>
          </Menu.List.Item>
        </Menu.List>
      </Menu>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
