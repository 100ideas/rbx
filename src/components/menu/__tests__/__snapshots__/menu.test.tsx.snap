// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component should Exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "List": Object {
    "$$typeof": Symbol(react.forward_ref),
    "Item": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "active": false,
        "children": null,
      },
      "render": [Function],
    },
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "aside",
  },
  "render": [Function],
}
`;

exports[`Menu component should concat Menu.List to display as submenus 1`] = `
<aside
  className="menu"
>
  <p
    className="menu-label"
  >
    General
  </p>
  <ul
    className="menu-list"
  >
    <li>
      <a>
        Dashboard
      </a>
    </li>
    <li>
      <li>
        <a
          className="is-active"
        >
          Manage Your Team
        </a>
        <ul
          className="menu-list"
        >
          <li>
            <a>
              Members
            </a>
          </li>
          <li>
            <a
              className="is-active"
            >
              Plugins
            </a>
          </li>
          <li>
            <a>
              Add a member
            </a>
          </li>
        </ul>
      </li>
    </li>
  </ul>
</aside>
`;

exports[`Menu component should have menu classnames 1`] = `
<aside
  className="menu"
>
  <p
    className="menu-label"
  >
    General
  </p>
  <ul
    className="menu-list"
  >
    <li>
      <a>
        Dashboard
      </a>
    </li>
    <li>
      <a>
        Customer
      </a>
    </li>
  </ul>
</aside>
`;

exports[`Menu component should render custom item child 1`] = `
<aside
  className="menu"
>
  <p
    className="menu-label"
  >
    General
  </p>
  <ul
    className="menu-list"
  >
    <li>
      <p>
        Custom children 1
      </p>
    </li>
    <li>
      <a>
        Custom children 2
      </a>
    </li>
  </ul>
</aside>
`;
